streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -473,5 +485,45 @@ public StreamingMetrics() {
             this.taskDestructionSensor = metrics.sensor("task-destruction");
             this.taskDestructionSensor.add(new MetricName("task-destruction-rate", metricGrpName, "The average per-second number of destructed tasks", metricTags), new Rate(new Count()));
         }
+
+        @Override
+        public void recordLatency(Sensor sensor, long startNs, long endNs) {
+            sensor.record((endNs - startNs) / 1000000, endNs);
+        }
+
+        @Override
+        public Sensor addLatencySensor(String scopeName, String entityName, String operationName, String... tags) {
+            // extract the additional tags if there are any
+            Map<String, String> tagMap = new HashMap<>(this.metricTags);
+            if ((tags.length % 2) != 0)
+                throw new IllegalArgumentException("Tags needs to be specified in key-value pairs");
+
+            for (int i = 0; i < tags.length / 2; i++)
```

Comment:
@guozhangwang Forgive me if I understood wrong, but this method, really. But if the list is composed of key-value pairs, wouldn't this for loop goes up to the the middle of the array, no? Plus, if ``tags`` is ``"a","1","b","2"`` the final map will be `` "a" -> "1", "1" -> "b"``. Is this the right intend? If not, I suppose we could rewrite it as 

```
for (int i = 0; i < tags.length; ) {
     tagMap.put(tags[i], tags[i+1]);
     i += 2;
}
```

Again, forgive me if I understood everything wrong.